### ISSUE TITLE
Graphene Evaluation

### ISSUE THREAD

**Issue by sdrunima** [Lambda-3/Graphene#31]
Hi,

I recently worked on a modified Graphene implementation and tried to replicate the evaluation in the Graphene paper. However, I have some questions regarding your evaluation.
1) I'm wondering which confidence values you provided to the benchmark, as it seems like Graphene's default RE initializes the confidence of all extractions with null and never changes these values.

2) Is the code for your evaluation available or could you provide more information about it, especially regarding the matching of gold extractions with Graphene's extractions? Also, how did you identify the grammatical head of the rel/args. If I understand the description in the Graphene paper correctly, your implementation is different than the default implementation of the oie-benchmark.

Thank you in advance!

**Comment by Mischn** [comment]
Hello,

I am glad to see your interest in Graphene.
Our evaluation code is not publicly available, but I will explain some of our evaluations settings in more detail now:

In order to generate the Graphene output, we used the following Graphene-setting (set in Graphene's _reference.conf_):

```
    relation-extraction {
		relation-extractor = org.lambda3.graphene.core.relation_extraction.impl.NestedRelationExtractor

		exploit-core = true
		exploit-contexts = false
		separate-noun-based = true
		separate-purposes = true
		separate-attributions = true
    }
```


At the moment, Graphene does not calculate confidence scores when processing text. So the calculation of the confidence scores was done afterwards in Python using the following function (_text_ denotes the input sentence):

```
def calc_confidence(self, text):
        return 1./(0.005*len(text)+1)
```

We indeed changed the default implementation of the framework to the algorithm described by Stanovsky and Dagan (2016) where extractions are compared with their gold references by head-word-matching. For identifying head-words, we use the root of the dependency-parsed output from the StanfordDependencyParser (using Python's _nltk.parse.stanford_ package):

```
DEP_PARSER = StanfordDependencyParser(path_to_jar=..., path_to_models_jar=...)

def dep_parse(text):
    return list(DEP_PARSER.raw_parse(text))[0]

def dep_head(graph):
    return graph.root['word']

# Example:
graph = dep_parse("A text")
print dep_head(graph)
```

Here is a pseudo-code for the matching algorithm:

```
def headMatch(ref, ex, ignoreStopwords, ignoreCase):
        # parsing of head verbs
        refHeadPred = dep_head(ref.pred)
        exHeadPred = dep_head(ex.pred)
        refHeadArgs = [dep_head(a) for a in ref.args]
        exHeadArgs = [dep_head(a) for a in ex.args]

        predmatch = True
        argmatch = True

        # match pred (head)
        if not refHeadPred == exHeadPred:
            predmatch = False

        # match arguments (head)
        matches = 0
        for refHeadArg in refHeadArgs:
            for exHeadArg in exHeadArgs:
                if refHeadArg == exHeadArg:
                    matches += 1
                    break
        if matches < 2:
            argmatch = False

        return predmatch and argmatch
```

**Comment by sdrunima** [comment]
Thank you for the detailed answer Matthias!
I think everything is clear now. This really helps a lot with the evaluation.

**Comment by Ahmedn1** [comment]
Can you explain the confidence calculation to me more? 
I mean it seems that your functions calculates confidence based only on input length. That doesn't relate to the confidence in each extraction!

**Comment by Aatlantise** [comment]
Another follow-up question; does the len(text) indicate word length or char length? Just to be clear.

Much thanks!

**Comment by Mischn** [comment]
Hello,

Since Graphene does not produce confidence-scores for its extractions, we needed to find a quick way how to generate "pseudo" confidence-scores in order to integrate our system into the [Stanovski benchmark](https://github.com/gabrielStanovsky/oie-benchmark).

Because Graphene's performance highly depends on the performance of the lexicalized parser that is being used, we chose a simple heuristic that is based on the overall input size of the sentence, assuming that longer input sentences lead to more parsing errors.

So the chosen heuristic simply uses the character-length of the input-sentence (as described above) for all extractions of an input sentence.

A better approach would be to utilize the confidence of the lexicalized parser and propagate it down to individual extractions, but due to lack of time, we did not implement this.
